### ISSUE TITLE
Flash attention on cc>=7.5 is broken by the CUDA 11.4 toolchain — decide: leave off, route to WMMA, or root-cause

### ISSUE THREAD

**Issue by dogkeeper886** [dogkeeper886/ollama37#388]
## Description

Flash attention — the speed trick behind text generation — produces garbage on every card newer
than a Volta. #385 fixes the *symptom* by switching it off. This issue is about the *cause*, and
about deciding whether those cards should ever get it back.

The reason is a chain of consequences we cannot easily break. The K80 is old, so we must build
with an old NVIDIA compiler (CUDA 11.4); newer compilers dropped the K80 entirely. That old
compiler lacks one small instruction the fast attention code wants. When it's absent, the code
quietly substitutes a hand-written stand-in. On the K80 the stand-in is never used, because the
K80 never runs this code. On a 2060 it *is* used — and the numbers come out wrong.

**The card that forces us onto the old compiler is not the card that gets hurt by it.**

## Expected Behavior

Either flash attention works on cc >= 7.5, or we state plainly and permanently that it does not,
with the reason recorded. What we must not have is a feature that silently enables itself and
produces nonsense.

## Goal

Decide, on evidence, what flash attention does on cards newer than the K80 — and if it can be made
to work cheaply, make it work; otherwise close the door and document why.

## Steps to Reproduce

Same as #385, with FA forced on to remove any doubt:

1. cc >= 7.5 host (RTX 2060 here), published image, `ollama-data` volume created.
2. `printf 'OLLAMA_FLASH_ATTENTION=1\n' > docker/.env && cd docker && docker compose up -d --force-recreate`
3. `docker exec ollama37 ollama pull deepseek-r1:1.5b`
4. `curl -s localhost:11434/api/generate -d '{"model":"deepseek-r1:1.5b","prompt":"In one sentence, what is the capital of France?","stream":false,"options":{"num_predict":16,"num_ctx":4096,"seed":42}}'`

**Result (2/2 runs):** `llama-sampling.cpp:662: llama_sampler_dist_apply: Assertion 'found' failed`
→ `SIGABRT`.

**Control (passes):** same command with `CUDA_VISIBLE_DEVICES=-1` → `library=cpu`, FA genuinely
enabled (`"enabling flash attention"`), 4.77 tok/s, no abort. **The fault is in the CUDA branch.**

## Environment

| | |
|---|---|
| GPU | RTX 2060 (TU106), **cc 7.5**, driver 580.159.03 |
| Image | `dogkeeper886/ollama37:latest` `sha256:32010cd31c32…` (`2.1.0`) |
| Bundled CUDA runtime | `libcudart.so.11.4.148` ⇒ `CUDART_VERSION` = **11040** |
| Builder | `docker/builder/Dockerfile` — CUDA 11.4, GCC 10 |

## The path

**Observed** — the chain to the failure:

```
fattn.cu:195   ggml_cuda_get_best_fattn_kernel
                 turing_mma_available(750) → TRUE          (common.cuh:317)
fattn.cu:295   → BEST_FATTN_KERNEL_MMA_F16
                 → _switch_ncols2 → _switch_ncols1 → _mma_f16_case
fattn-mma-f16.cuh:696,1142   get_transposed(...)
mma.cuh:213                  get_transposed → ggml_cuda_movmatrix ×2
mma.cuh:22                   #if CUDART_VERSION >= 11080  →  11040, FALSE
mma.cuh:39                   → software __shfl_sync fallback     ← SUSPECT, not proven
                             ↓  NaN / all-zero logits
                             sampler aborts (both engines)
```

`ggml_cuda_movmatrix` has exactly one consumer, `get_transposed`, called from exactly two lines,
**both in `fattn-mma-f16.cuh`**. Nothing else in the backend calls it. On the K80 the code compiles
but never executes — `turing_mma_available(370)` is false. That is why it was never seen.

## Blast radius

| Card | cc | FA kernel | Affected |
|---|--:|---|---|
| K80 | 3.7 | none — `ml/device.go:436` vetoes FA | no |
| V100 | 7.0 | WMMA (`== GGML_CUDA_CC_VOLTA`) | **likely not — untested** |
| T4 / RTX 20xx | 7.5 | `MMA_F16` | **yes — observed** |
| A100 / A40 / RTX 30xx | 8.0 / 8.6 | `MMA_F16` | **yes — predicted** |

`GGML_CUDA_CC_TURING` is 750, so the boundary is **cc >= 7.5**, not 7.0.

## Candidate fixes

**A. Leave it off (status quo after #385).** Zero risk, zero cost. Those cards lose a feature they
never had here. The K80 is unaffected either way.

**B. Route cc >= 7.5 to the WMMA kernel.** The WMMA flash-attention kernels are **already compiled
into our sm_75/80/86 cubins** — `fattn-wmma-f16.cuh:5` defines `GGML_USE_WMMA_FATTN` for
`__CUDA_ARCH__ >= GGML_CUDA_CC_VOLTA`, and `fattn-wmma-f16.cu` includes nvcuda's `<mma.h>`, **not**
our `mma.cuh`. It never touches `get_transposed`. They are dead code today.

Two gates keep those cards away from them, so this is **not** a one-line change:

```
fattn.cu:273              if (turing_mma_available(cc) && …)  → returns MMA_F16 first
fattn-wmma-f16.cuh:30     ggml_cuda_highest_compiled_arch(cc) == GGML_CUDA_CC_VOLTA
                                                              ↑ EXACT equality, Volta only
```

Costs and risks, stated honestly:
- Edits vendored upstream ggml in two places — a fork-hack, which #340's guiding principle warns
  against and which the next re-sync must re-apply.
- Upstream restricted WMMA to Volta **deliberately** once MMA landed, presumably on performance.
  Per `docs/porting-k80.md` §2, this must be **benchmarked**, not assumed. This project has been
  burned exactly here before (FA on the K80: assumed ~3% win, actually a 7× loss).
- The two engines would not take the same sub-path: the Go engine forces `GGML_PREC_F32`
  (`ml/backend/ggml/ggml.go:1805`) and `ggml_cuda_flash_attn_ext_wmma_f16` branches on precision.
- The K80 is untouched — the device guard is `__CUDA_ARCH__ >= VOLTA`, so sm_37 cubins contain none
  of it.

**C. Confirm the root cause first.** A `get_transposed` correctness probe compiled for sm_75 in the
builder image, or an sm_75-only CUDA 11.8 build (which would drop sm_37 and cannot ship).

## Non-goals

- **Do not raise the CUDA version.** CUDA >= 11.8 gives the missing instruction but drops Kepler.
  sm_37 is why this fork exists (`docs/porting-k80.md` §1: port the source, don't adapt the
  environment).
- Do not extend K80 fork-hacks to the new cards (#340's stated principle) without a benchmark.

## Acceptance criteria

- [ ] A decision is recorded: A, B, or C — with the reasoning, in this issue.
- [ ] If **B**: `qwen3.5:0.8b` and `deepseek-r1:1.5b` run with FA on at cc 7.5 with correct output,
      **and** measured tok/s is compared against FA-off. Keep only if it wins.
- [ ] **K80 no-harm:** the sweep is unchanged, ±3% tok/s across all model × ctx rows
      (`docs/reports/k80-vram-by-family-2026-07-10-d83061de.md`). Any option here must be a no-op
      on sm_37.
- [ ] The outcome is documented wherever the widened build is described.

## Unverified

- **The `movmatrix` fallback is a suspect, not a confirmed defect.** T6 (CPU control) localized the
  fault to the CUDA FA path; it did **not** single out `get_transposed` within it. The fault could
  be elsewhere in `fattn-mma-f16` under nvcc 11.4.
- **V100 (sm_70) has never been tested.** It is *predicted* to be unaffected because it takes WMMA.
  No Volta hardware available.
- **Option B has never been compiled**, let alone benchmarked.

## Links

- Symptom + mitigation: #385
- Evidence, full matrix (T1–T6): https://github.com/dogkeeper886/ollama37/issues/342#issuecomment-4934523329
- Plan and host state: `docs/research/sm75-fa-validation-plan.md`
- Amends #340 step 2, which assumes cc >= 7 cards get "pure-upstream and already correct" FA.

**Comment by dogkeeper886** [comment]
## Superseded by #385

This issue was framed as a decision — *leave FA off, route cc>=7.5 to WMMA, or root-cause it.* Tracing the selector removed the decision.

**The WMMA option is dead.** It required two edits, not one: bypassing the MMA branch at `fattn.cu:273` **and** relaxing `ggml_cuda_should_use_wmma_fattn`'s `== GGML_CUDA_CC_VOLTA` (exact equality, `fattn-wmma-f16.cuh:30`) to a `>=`. Both in vendored upstream ggml.

**It is unnecessary.** Skipping `MMA_F16` already falls through to `BEST_FATTN_KERNEL_TILE` (`fattn.cu:320`) — a kernel with no `mma.cuh` dependency that covers head dim 128 (`fattn-tile.cu:29`). One predicate, no equality change, no WMMA involvement.

**And "leave it off" was never a fix**, only a mitigation — and PR #393 showed the mitigation didn't even cover the `docker run` path the README documents.

The real defect: `ggml_cuda_get_best_fattn_kernel` selects `MMA_F16` on a **hardware** predicate while that kernel carries an unstated **toolkit** precondition (`movmatrix.sync.aligned` → PTX ISA 7.8 → CUDA 11.8). The fix is a `ggml_cuda_should_use_mma_fattn()` predicate in `fattn-mma-f16.cuh`, mirroring upstream's own WMMA idiom.

Everything from this issue — the blast-radius table (cc>=7.5, not 7.0), the CPU control, the `movmatrix` trace, the K80 byte-identical argument, the `porting-k80.md` §2 benchmark requirement — is carried into #385, which is now the fix issue.

Two things preserved here that are still true and still unproven:
- The `movmatrix` fallback is a **suspect, not a confirmed defect**. The fix is correct either way.
- **sm_70 (Volta) has never been tested.** It takes the WMMA kernel and is predicted unaffected. No Volta hardware.

Closing. Work continues in #385.
